packages/explat-client/src/create-explat-client.ts

```diff
@@ -0,0 +1,178 @@
+/**
+ * Internal dependencies
+ */
+import { ExperimentAssignment, Config } from './types';
+import * as ExperimentAssignments from './internal/experiment-assignments';
+import * as Request from './internal/requests';
+import * as State from './internal/state';
+import * as Timing from './internal/timing';
+import * as Validation from './internal/validations';
+import { createFallbackExperimentAssignment as createFallbackExperimentAssignment } from './internal/experiment-assignments';
+
+/**
+ * The number of milliseconds before we abandon fetching an experiment
+ */
+const EXPERIMENT_FETCH_TIMEOUT = 5000;
+
+export interface ExPlatClient {
+	/**
+	 * Loads and returns an Experiment Assignment Promise, starting an assignment if necessary.
+	 *
+	 * Call as many times as you like, it will only make one request at a time (per experiment) and
+	 * will only trigger a request when the assignment TTL is expired.
+	 *
+	 * Will never throw in production, it will return the default assignment.
+	 * It should not be run on the server but it won't crash anything.
+	 *
+	 * @param experimentName The experiment's name
+	 */
+	loadExperimentAssignment: ( experimentName: string ) => Promise< ExperimentAssignment >;
+
+	/**
+	 * Get an already loaded Experiment Assignment, will throw if there is an error, e.g. if it hasn't been loaded.
+	 *
+	 * Make sure loadExperimentAssignment has been called before calling this function.
+	 *
+	 */
+	dangerouslyGetExperimentAssignment: ( experimentName: string ) => ExperimentAssignment;
+}
+
+export class MissingExperimentAssignmentError extends Error {
+	constructor( ...params: any[] ) {
+		super( ...params );
+
+		// Maintains proper stack trace for where our error was thrown (only available on V8)
+		if ( Error.captureStackTrace ) {
+			Error.captureStackTrace( this, MissingExperimentAssignmentError );
+		}
+
+		this.name = 'MissingExperimentAssignmentError';
+	}
+}
+
+/**
+ * Create an ExPlat Client
+ *
+ * @param config Configuration object
+ */
+export default function createExPlatClient( config: Config ): ExPlatClient {
+	/**
+	 * This bit of code is the heavy lifting behind loadExperimentAssignment, allowing it to be used intuitively.
+	 *
+	 * AOAAT stands for Async One At A Time, this is how we ensure for each experiment that there is only ever one fetch process occuring.
+	 *
+	 * @param experimentName The experiment's name
+	 */
+	const createAOAATExperimentAssignmentFetchAndStore = ( experimentName: string ) =>
+		Timing.asyncOneAtATime( async () => {
+			const fetchedExperimentAssignment = await Request.fetchExperimentAssignment(
+				config,
+				experimentName
+			);
+			State.storeExperimentAssignment( fetchedExperimentAssignment );
+			return fetchedExperimentAssignment;
```

Comment:
@withinboredom RE: [extending the async-one-at-a-time over an interval after it fetches](https://github.com/Automattic/wp-calypso/pull/48475#discussion_r567688124)

> I mean, there's a brief moment between when this function completes and the result is stored in memory. It'd be good to hold on to the promise for a bit longer to give it time to commit.

The fetch and store both happen together wrapped within the `asyncOneAtATime` - either the promise has resolved and it has successfully stored and future calls will retrieve it, the promise has rejected and we have (hopefully) stored a fallbackExperimentAssignment with a 60s TTL, or the promise is still pending and future calls will return the ExperimentAssignment.

In my mind we should be airtight here.

Reply:
We should possibly be wrapping the store of `fallbackExperimentAssignment` but it practice it is fine thanks to retrieving the "stale" experiment assignment.

Reply:
Yep, sorry, I should have mentioned that it is fine now :) I was just clarifying my comment!